### PR TITLE
Use 'name' from plugin descriptor file to determine plugin name

### PR DESCRIPTION
**Pull request by cbuescher**
At the moment, when installing from an url, a user provides a plugin name on the command line like:
- bin/plugin install [plugin-name] --url [url]

This can lead to confusion when picking an already existing name from another plugin, which might even overwrite plugins already installed with that name.

In order to remedy this, this PR introduces an additional mandatory `name` property to the plugin descriptor file which overwrites the name provided by the user after the plugin was loaded and the descriptor file is unpacked.

Relates to #12715

### CODE REVIEW

**Comment by cbuescher** [comment]
Some open questions I have with the original issue which are not addressed by this PR so far:

* What should happen with the `[plugin-name]` CLI argument. We can't remove it, since it is used and needed when no url is provided. As far as I understand, in this case the name is going to be resolved from the CLI argument depending on various rules.
* When the name is specified in the descriptor file, I think that also that name needs to be used to remove the plugin. Not sure how much this is in conflict with existing behaviour

File: core/src/main/java/org/elasticsearch/plugins/PluginInfo.java [hunk 1]
```diff
@@ -95,6 +95,10 @@ public static PluginInfo readFromProperties(Path dir) throws IOException {
         if (version == null) {
             throw new IllegalArgumentException("Property [version] is missing for plugin [" + name + "]");
         }
+        String pluginName = props.getProperty("plugin.name");
+        if (pluginName == null) {
+            throw new IllegalArgumentException("Property [plugin.name] is missing for plugin [" + name + "]");
+        }
```
Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: I'm wondering if in that case we should not fallback to the previous convention (use zip name as the plugin name) and just "warn" the user?

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: Yes, I wasn't too sure about the original issues goals here, so I didn't elaborate any further.

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: I think the whole point of this PR is to add this new parameter and make it required, so I don't think we should fall back? Otherwise the leniency that we are trying to fix is still here.

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: In particular, since we now require that plugins have exactly the same version as elasticsearch, plugins will have to be re-packaged anyway, so I don't think bw compat is a concern?

**Comment by dadoonet** [comment]
Some thoughts that come to mind. 

Now we know that if the user set a `url` like:

```
bin/plugin install foo --url http://link.to/myplugin.zip
```

And that `myplugin.zip` contains in its descriptor a `plugin.name` equal to `bar`, we will end up installing `bar` plugin.

So it does not make sense anymore to support this `--url` option IMO.

We should have:

```sh
# download from elastic.co, maven central or github
bin/plugin install foo
# download from a URI (URL)
bin/plugin install http://link.to/foo.zip
# download from a URI (file)
bin/plugin install file:.target/releases/foo.zip
```

@spinscale @clintongormley WDYT?

File: core/src/main/java/org/elasticsearch/plugins/PluginInfo.java (hunk 1 above)
Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: Not a big deal. I was just thinking of community plugin developers but it's not hard for them to fix. We should may be add this in BWC doc.

**Comment by rjernst** [comment]
Hrm, @dadoonet Can we be explicit instead of having to infer *what* was passed as the argument to install? What you suggest would require even more heuristics. Instead, why don't we remove the fallback to trying as a url, change http:// to require using --url, and add a --file that requires a local path?

File: core/src/main/java/org/elasticsearch/plugins/PluginInfo.java [hunk 2]
```diff
@@ -95,6 +95,10 @@ public static PluginInfo readFromProperties(Path dir) throws IOException {
         if (version == null) {
             throw new IllegalArgumentException("Property [version] is missing for plugin [" + name + "]");
         }
+        String pluginName = props.getProperty("plugin.name");
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: Hrm, can we just call this "name"? None of the other settings are prefixed with plugin.

**Comment by rjernst** [comment]
Also, when we download a plugin, if it was downloaded by name, we should verify the name matches?

File: core/src/main/java/org/elasticsearch/plugins/PluginInfo.java (hunk 1 above)
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: This file is new in 2.0, we can change it

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: Also, can you add an element to maven enforcer plugin for plugins/pom.xml so it fails build cleanly and early if this property is not set? We should also insert a check in pluginservice, if it differs from the directory name, someone manually meddled

Comment (rjernst) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: We should also check the name matches that in jvm plugins. As a follow up, we should at least remove description from jvm Plugin interface, and possibly also name (possibly a little harder, just requires passing around PluginInfo instead of Plugin I think).

Comment (rjernst) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: Actually I just checked and removing name and description from the Plugin interface should be easy. The only thing to think about is what to give for those properties when plugins are loaded from the classpath (plugin.types). I think here the name should just be the classname, and description something like "plugin loaded from classpath"? I don't know what other info we really have.

File: core/src/main/java/org/elasticsearch/plugins/PluginInfo.java (hunk 2 above)
Comment (dadoonet) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: +1

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: Sure, changed the property to just 'name'.

File: core/src/main/java/org/elasticsearch/plugins/PluginInfo.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: @rmuir added element to the enforcer plugin, but I'm not sure what check to insert in pluginservice and how that service actually works. Can this be added in a follow up PR?

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: @rjernst Ryan, not sure what you mean by jvm plugins and the checks your suggested, can this be done outside of this PR? Otherwise where do I need to look?

**Comment by clintongormley** [comment]
I agree with being explicit:

    # download from elastic.co, maven central or github
    bin/plugin install foo
    # download from a URI (URL)
    bin/plugin install --url http://link.to/foo.zip
    # download from a URI (file)
    bin/plugin install --file .target/releases/foo.zip

**Comment by dadoonet** [comment]
@spinscale Does CLI allows that?

I think we might need to do something like:

```sh
# download from elastic.co, maven central or github
bin/plugin install foo
# download from a URI (URL)
bin/plugin install http://link.to/foo.zip --url 
# download from a URI (file)
bin/plugin install .target/releases/foo.zip --file
```

**Comment by spinscale** [comment]
the last two syntaxes are really confusing IMO, one can change to allow for that, but it doesnt feel nice

**Comment by clintongormley** [comment]
If we can't easily support the version in https://github.com/elastic/elasticsearch/pull/12775#issuecomment-129937866 then let's use the version in https://github.com/elastic/elasticsearch/pull/12775#issuecomment-129519080

The heuristic would simply be: anything matching this regex is a URL `^\w+:`

**Comment by rjernst** [comment]
Wasn't @spinscale talking about the odd syntax of adding options after the thing the option is pointing to? Surely having command line parameter that takes an argument is supported by the CliTool? We should simply not use heuristics; they are not needed, and only lead to confusing errors.

**Comment by clintongormley** [comment]
@rjernst yes he was.  @dadoonet was not sure if the CLI parser we use would support that style of argument parsing.  I'm saying that, if it doesn't (and is not easy to fix) then we should rather use the simple heuristic than the confusing syntax.

But I agree, I can't see why the CLI parser wouldn't support it.

**Comment by spinscale** [comment]
Looks good me to regarding support... One last thing though:

```
# downloads official plugin
bin/plugin install analysis-kuromoji 
# downloads github plugin
bin/plugin install lmenezes/elasticsearch-kopf
# URL
bin/plugin install http://link.to/foo.zip
bin/plugin install file://.target/releases/foo.zip
# local
bin/plugin install --file .target/releases/foo.zip
```

Do we really need a `--url` and `--file` switch? How about just providing an URL by default, which also works if you provide `file://`. As we try to remove leniency, rather than add it at the moment, we could still have the `--file` switch, which automatically prepends `file://` to the specified path to create an URI instead of trying to be smart and fall back to do this automatically if parsing as an URL fails.

**Comment by cbuescher** [comment]
@spinscale Thanks for that suggestions, looking at the implementation details and given that `-u` currently also covers installation from file I'd also prefer not adding yet another switch. Will try this first without the `--file` switch.

**Comment by rjernst** [comment]
@spinscale I like this approach. In this case we shoudln't provide `--file` at all (we didn't before right? we only had `--url`?).

So to confirm, the logic would be:
```
if (argument-parses-as-uri) {
  ...use URI
} else {
  ...use as name with download service
}
```

**Comment by rjernst** [comment]
But that means we should remove `--url` too, otherwise we still would have the issue of the user passing a name with a url.

**Comment by spinscale** [comment]
@rjernst depening on our impl, logic would be

```
if (argument-parses-as-uri) {
  ...use URI
} else {
   if (argument-parses-as-uri-with-file://-prefix) {
  } else { 
    ...use as name with download service
  }
}
```

and yes, we could totally drop the --url in that case and just have a single argument, the identifier, being a name, local path or URL

**Comment by rjernst** [comment]
Why would the inner if be necessary? `file://` should parse as a url right?

**Comment by spinscale** [comment]
@rjernst sure, but you might want to use paths like `/my/path/to/my/plugin.zip` without specifying file:// in front

**Comment by rjernst** [comment]
We should not do that. That is leniency. Using a local file is expert, let's just stick to a uri, and not have to deal with `is-this-thing-a-file-or-should-we-try-the-download-service`

**Comment by rjernst** [comment]
It's also no different than what users already had to do, using --url. We can just remove --url, since we can always clearly tell if something is a uri.

**Comment by spinscale** [comment]
valid argument, +1

**Comment by cbuescher** [comment]
@rjernst @spinscale updated the PR with changes to the CLI as mentioned in https://github.com/elastic/elasticsearch/pull/12775#issuecomment-130214795, only without the `--file` option, also removing the `--url` option. I choose not to try to parse to an URI since also valid plugin names parse to URIs, only without a scheme, let me know if you think this is okay. Also, if only url/file is supllied, I seem to need a temporary name for the download, not sure what to do here. I mentioned this in a comment.

**Comment by rjernst** [comment]
> Also, if only url/file is supllied, I seem to need a temporary name for the download

How about using part of the URL in this case? Maybe the path?

File: core/src/main/resources/org/elasticsearch/plugins/plugin-install.help [hunk 3]
```diff
@@ -20,6 +20,8 @@ DESCRIPTION
 
     The notation of 'username/repository' refers to a github repository.
 
+    The argument can be an valid <url> which points to a download or file location for the plugin to be loaded from.
+
 EXAMPLES
 
     plugin install elasticsearch-analysis-kuromoji
```
Comment (spinscale) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: drive-by-fix: this should be `analysis-kuromoji`

Comment (cbuescher) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: So is the es prefix wrong now? Asking because there is a list with supported plugin names in further down in this help file, should I update those names too?

Comment (rjernst) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: I think we should remove the `elasticsearch-` prefix from those in that list.

Comment (cbuescher) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: @rjernst @spinscale looks like if I touch that list and remove all `elasticsearch-` prefixes there, a PluginManagerTests that checks the help output vs. the `OFFICIAL_PLUGINS` list in the PluginManger breaks. I could also change that, but I'm not sure if I'm going down a rabbit hole with this. I think this should to be done in a consistent way in an additional PR.

Comment (dadoonet) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: I think that we should rename our artifacts. I know it's late but if we want to be consistent we should remove elasticsearch- from all plugins

Comment (rjernst) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: +1!!!!

Comment (cbuescher) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: @dadoonet Again, can this be done in a PR that focuses on that part? I would prefer not to mix too many moving parts here.

Comment (dadoonet) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: If your change is not related to the issue @spinscale [found](https://github.com/elastic/elasticsearch/pull/12775/files#r37058692), then yes we should fix that in another issue.

Comment (cbuescher) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: @dadoonet no I think it is not related, since you are probably more familiar with the implications of changing the artifact names and what else needs to be done, could help with formulating that follow up issue?

File: core/src/main/java/org/elasticsearch/plugins/PluginInfo.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: @rjernst since you already seem to have some additional checks in mind, could you open that in a follow up issue or let me know what really needs to be done in this PR.

**Comment by cbuescher** [comment]
Rebased and added some randomness to the temporary plugin name that is used when url is provided but descriptor file hasn'r been retrieved yet. I tried getting around this but this would have meant larger rewrites of the plugin manager code and as far I understand @spinscale this will happen at a later point in time anyway.

File: dev-tools/src/main/resources/ant/integration-tests.xml [hunk 4]
```diff
@@ -275,7 +271,7 @@
 
   <!-- unzip core release artifact, install plugin, then start ES -->
   <target name="start-external-cluster-with-plugin" depends="setup-workspace">
-    <install-plugin name="${project.artifactId}" file="${project.build.directory}/releases/${project.artifactId}-${project.version}.zip"/>
+    <install-plugin name="${project.artifactId}" fileUrl="file://${project.build.directory}/releases/${project.artifactId}-${project.version}.zip"/>
```
Comment (rmuir) [inline dev-tools/src/main/resources/ant/integration-tests.xml]: This isn't correct. Why make this change? the call to makeurl as it is done currently in master is the only correct way. string concatenation simply will not work, if the directory contains whitespace for example, or certain other characters (it must be a true URL)

File: core/src/main/resources/org/elasticsearch/plugins/plugin-install.help (hunk 3 above)
Comment (dadoonet) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: @cbuescher Yeah. I discussed about it with @rjernst and I'm working at the moment on a PR

File: dev-tools/src/main/resources/ant/integration-tests.xml (hunk 4 above)
Comment (cbuescher) [inline dev-tools/src/main/resources/ant/integration-tests.xml]: Thanks, reverted that change.

File: core/src/main/java/org/elasticsearch/plugins/PluginManagerCliParser.java [hunk 5]
```diff
@@ -166,19 +168,28 @@ public ExitStatus execute(Settings settings, Environment env) throws Exception {
...
-
             String name = args[0];
+
+            URL pluginUrl = null;
+            // try provided name as URL,
+            try {
+                pluginUrl = new URL(name);
+                name = null;
+            } catch (MalformedURLException e) {
+                // ignore
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/plugins/PluginManagerCliParser.java]: This seems strange - maybe explain why its ok to ignore this here?

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/plugins/PluginManagerCliParser.java]: If we do `plugin install [nameOrUrl] and the argument doesn't parse to a valid url, we go on treating it as on of the other three options of a plugin name (e.g. analysis-icu, user/repo etc...). I can update the comment to make this clearer.

File: core/src/main/java/org/elasticsearch/plugins/PluginInfo.java [hunk 6]
```diff
@@ -95,6 +95,12 @@ public static PluginInfo readFromProperties(Path dir) throws IOException {
         if (version == null) {
             throw new IllegalArgumentException("Property [version] is missing for plugin [" + name + "]");
         }
+        String pluginName = props.getProperty("name");
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: Now the logging here is really "fun". Maybe pull the pluginName to the top and use it in the []s. I'm trying to think through what this will look like for the temporary directory versions. Maybe it makes sense to make the whole temporary directory thing a more first class citizen - like a field or a subclass of this rather than just passing in randomness.

File: core/src/main/java/org/elasticsearch/plugins/PluginManagerCliParser.java (hunk 5 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/plugins/PluginManagerCliParser.java]: Ah ok - yeah. Please update the comment. It'll help people just scanning the code.

File: core/src/main/java/org/elasticsearch/plugins/PluginInfo.java (hunk 6 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/plugins/PluginInfo.java]: Will do, I was not going deeper into the temp dir thing because this would change a lot in the plugin manager and I was told it will get a real update sooner or later anyway.

**Comment by dadoonet** [comment]
@cbuescher I think this file needs also to be updated with your change: https://github.com/elastic/elasticsearch/blob/master/docs/plugins/plugin-script.asciidoc#custom-url-or-file-system

**Comment by cbuescher** [comment]
Updated the PR, rebased and added comments, changes in logging suggested by @nik9000 and changed docs.

File: core/src/main/resources/org/elasticsearch/plugins/plugin-install.help [hunk 7]
```diff
@@ -20,14 +20,20 @@ DESCRIPTION
...
-    plugin install elasticsearch-analysis-kuromoji
+    plugin install analysis-kuromoji
 
     plugin install elasticsearch/shield/latest
 
     plugin install lmenezes/elasticsearch-kopf
 
+    plugin install http://download.elasticsearch.org/elasticsearch/elasticsearch-analysis-kuromoji/elasticsearch-analysis-kuromoji-2.7.0.zip
+
+    plugin install file:///path/to/plugin/elasticsearch-analysis-kuromoji-2.7.0.zip
```
Comment (dadoonet) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: Sometimes non java devs are confused by that and I don't think `//` is required. May be this?

```
      plugin install file:/path/to/plugin/elasticsearch-analysis-kuromoji-2.7.0.zip
or
      plugin install file:./path/to/plugin/elasticsearch-analysis-kuromoji-2.7.0.zip
```

Or may be adding a windows user example like this (not tested) ?

```
      plugin install file:c:/path/to/plugin/elasticsearch-analysis-kuromoji-2.7.0.zip
```

**Comment by dadoonet** [comment]
I left a small comment but the change looks good to me.

File: core/src/main/resources/org/elasticsearch/plugins/plugin-install.help (hunk 7 above)
Comment (cbuescher) [inline core/src/main/resources/org/elasticsearch/plugins/plugin-install.help]: Thanks for the suggestions, will test this.

**Comment by dadoonet** [comment]
@cbuescher I merged #12879 so you can now change pom.xml files to use maven artifactId instead of setting manually `elasticsearch.plugin.name`.

**Comment by cbuescher** [comment]
@dadoonet thanks, I updated the poms using the artifactId instead of hardcoding the name, added some documentation and rebased. I'll sqash this again before merging with master.

**Comment by dadoonet** [comment]
I left one small comment. 

I think we are really close now!

**Comment by cbuescher** [comment]
@dadoonet thanks for the review, didn't know `artifactId` would be resolved in the submodules if placeholder is defined in the parent pom, but seems to work. I updated the PR accordingly.

**Comment by dadoonet** [comment]
LGTM!

**Comment by cbuescher** [comment]
@dadoonet Rebased and squashed, will merge this to master so it can be tested there before we might more this to the branch.

**Comment by cbuescher** [comment]
Also backported to 2.0 branch with d236192c49a8941525e201e01edc781e5fc13dae

**Comment by cbuescher** [comment]
@clintongormley should this be labeled `breaking` since it affects the CLI syntax and adds mandatory properties to the plugin descriptor file?

**Comment by nik9000** [comment]
Update #12768 ?
